// playcanvas/build/playcanvas.dbg/src/extras/gizmo/constants.js
/**
 * Local coordinate space.
 *
 * @category Gizmo
 */ const GIZMOSPACE_LOCAL = 'local';
/**
 * World coordinate space.
 *
 * @category Gizmo
 */ const GIZMOSPACE_WORLD = 'world';
/**
 * Gizmo axis for the line X.
 *
 * @category Gizmo
 */ const GIZMOAXIS_X = 'x';
/**
 * Gizmo axis for the line Y.
 *
 * @category Gizmo
 */ const GIZMOAXIS_Y = 'y';
/**
 * Gizmo axis for the line Z.
 *
 * @category Gizmo
 */ const GIZMOAXIS_Z = 'z';
/**
 * Gizmo axis for the plane YZ.
 *
 * @category Gizmo
 */ const GIZMOAXIS_YZ = 'yz';
/**
 * Gizmo axis for the plane XZ.
 *
 * @category Gizmo
 */ const GIZMOAXIS_XZ = 'xz';
/**
 * Gizmo axis for the plane XY.
 *
 * @category Gizmo
 */ const GIZMOAXIS_XY = 'xy';
/**
 * Gizmo axis for all directions XYZ.
 *
 * @category Gizmo
 */ const GIZMOAXIS_XYZ = 'xyz';
/**
 * Gizmo axis for facing the camera (facing the camera).
 *
 * @category Gizmo
 */ const GIZMOAXIS_FACE = 'face';

export { GIZMOAXIS_FACE, GIZMOAXIS_X, GIZMOAXIS_XY, GIZMOAXIS_XYZ, GIZMOAXIS_XZ, GIZMOAXIS_Y, GIZMOAXIS_YZ, GIZMOAXIS_Z, GIZMOSPACE_LOCAL, GIZMOSPACE_WORLD };
